Mixed Appearance panel shows controls for valid layers in selection

DivyaPrabhakar opened the pull request:
Fixes issues #3777, #3756,  #3615, #3147, and #2457

Issue #3777:

Before:
![](https://cloud.githubusercontent.com/assets/1075154/13445927/20c345d2-dfc4-11e5-8ad6-827fb8e7b2be.png)

After: I have made all disabled inputs use the specified CSS color for each stop `@underlines-disabled`. In the examples below, blend mode and opacity are enabled and the rest of the type settings are disabled, just to give an idea of the different states relative to one another. @placegraphichere thoughts?

![](https://s3.amazonaws.com/f.cl.ly/items/291A2l3327052X3g1Z2r/Image%202016-03-04%20at%2011.26.36%20AM.png?v=c600a807)
![](https://s3.amazonaws.com/f.cl.ly/items/0h47451J1w0j2H352d3a/Image%202016-03-04%20at%2011.27.01%20AM.png?v=9763551d)
![](https://s3.amazonaws.com/f.cl.ly/items/1J0z3Z2u2y27122g0I3t/Image%202016-03-04%20at%2011.27.25%20AM.png?v=c675cdf9)
![](https://s3.amazonaws.com/f.cl.ly/items/1U0m2v0G3Z0h0P3p3R1J/Image%202016-03-04%20at%2011.27.51%20AM.png?v=1b9cf4f7)

Issue #3756

Opacity and Blend Mode Alignment
Before:
![](https://cloud.githubusercontent.com/assets/118264/13305420/f3dbf3e2-db0f-11e5-8a6b-95925d1b561e.png)

After:
![](https://s3.amazonaws.com/f.cl.ly/items/1g291t0F1E0N3R36143M/Image%202016-03-04%20at%2011.39.32%20AM.png?v=bb5c1dc7)

Color Picker Inputs Alignment
Before: 
![](https://cloud.githubusercontent.com/assets/118264/13305538/a55731b8-db10-11e5-979c-a96b3ad7cb8b.png)

After: 
![](https://www.dropbox.com/s/u4za3htzb3xmcvr/Screenshot%202016-03-04%2011.56.02.png?raw=1)


Radius Input Alignment: 
Before: 
![](https://cloud.githubusercontent.com/assets/118264/13305635/7e07588a-db11-11e5-863a-900a74a50db9.png)

After: 
![](https://www.dropbox.com/s/h6zud9zz4yws2l3/Screenshot%202016-03-04%2011.59.16.png?raw=1)


Issue #3615 
Before: 
![](https://camo.githubusercontent.com/b63e436db1cfc0cdf26a11275f625d6c109934b7/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f662e636c2e6c792f6974656d732f30643344335a324f3132336c33553367334b326f2f53637265656e2532305265636f7264696e67253230323031362d30312d3236253230617425323031322e3039253230504d2e6769663f763d6461383239623966)

After: 
![](https://s3.amazonaws.com/f.cl.ly/items/0l273v2Z3R2t293Y3A02/Screen%20Recording%202016-03-04%20at%2012.00%20PM.gif?v=c3402568)


Issue #3147 
Before: 
![](https://cloud.githubusercontent.com/assets/8442631/12793552/1063095c-ca66-11e5-9b5b-bd4b46528a29.png)

After: 
![](https://s3.amazonaws.com/f.cl.ly/items/0J3o052x0t0P3e0f061p/Screen%20Recording%202016-03-04%20at%2012.05%20PM.gif?v=72606480)

Issue #2457 @placegraphichere Can I get your opinion about this too? 

Before: 
![](https://cloud.githubusercontent.com/assets/1075154/9859811/3069171c-5add-11e5-938e-45e7745879d3.png)

After:
![](https://s3.amazonaws.com/f.cl.ly/items/0d091J0Y1c3B3H210r3F/Screen%20Recording%202016-03-04%20at%2012.12%20PM.gif?v=3d998291)

DivyaPrabhakar commented:
@volfied or @chadrolfs Can you make sure the now activated fields are the correct ones? Is there anything that should not be active? One thing I saw on `master` also is when two text layers are selected `Mixed Font Styles` is active but there are no options to choose from. Is that correct behavior?

ktaki commented:
Hi @DivyaPrabhakar Let me chime in. 

Found an issue. Create a rectangle and a text layer. Change the color of the rectangle. Select both layers. Now the color of rectangle in the panel is wrong.

![screen shot 2016-02-03 at 2 44 11 pm](https://cloud.githubusercontent.com/assets/2029597/12799679/fe8bae4a-ca84-11e5-9e76-199da9dab038.png)

When two text layers are selected, the current behavior looks right though more testing is needed.

![screen shot 2016-02-03 at 2 49 49 pm](https://cloud.githubusercontent.com/assets/2029597/12799747/6ee28222-ca85-11e5-8568-febc61f7e3bd.png)

chadrolfs commented:
A couple more things (leveraging Taki's screenshots above):
- When mixed state includes a shape and text, you can change the color of the text but not the shape
- In this mixed state ^^, blend and opacity are out of alignment
- There's no way that I can get the panel in to full view (without scrolling).  Collapsing Effects and Export collapse up

DivyaPrabhakar commented:
@ktaki and @chadrolfs 

1.) The appearance panel does not scroll anymore in the mixed state
![](https://www.dropbox.com/s/91hrht8nnodg31s/Screenshot%202016-02-05%2013.10.12.png?raw=1)

2.) The blend modes and opacity are now aligned 
![](https://www.dropbox.com/s/9blb5012clcwjj3/Screenshot%202016-02-05%2013.09.47.png?raw=1)

3.) The color picker for vector shapes is fixed
![](https://www.dropbox.com/s/k6j11tw90iud6s3/Screenshot%202016-02-05%2013.23.17.png?raw=1)
![](https://www.dropbox.com/s/1tjvhhzc778ucma/Screenshot%202016-02-05%2013.23.28.png?raw=1)

iwehrman commented on src/js/jsx/sections/style/VectorFill.jsx:
```diff
@@ -181,44 +181,52 @@ define(function (require, exports, module) {
                     layers={this.state.layers}
                     fill={this.state.fill} />);
+
+            if (this.props.document.layers.selected) {
```
This is always truthy. Maybe you want `!this.props.document.layers.selected.isEmpty()`?

iwehrman commented on src/js/jsx/sections/style/VectorFill.jsx:
```diff
@@ -181,44 +181,52 @@ define(function (require, exports, module) {
                     fill={this.state.fill} />);
+
+            if (this.props.document.layers.selected) {
+                var disabledVectorFill = this.props.document.layers.selected.filter(function (layer) {
```
This is legal and functionally correct (due to [variable hoisting](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/var#var_hoisting)), but for clarity I think we would prefer something like this:
```javascript
var disabledVectorFill;
if (this.props.document.layers.selected.isEmpty()) {
    disabledVectorFill = false;
} else {
    disabledVectorFill = // filter...
}
```

iwehrman commented:
It may well be inconvenient to implement this with our current component structure, but I definitely do not think showing a fill color swatch row AND a type color swatch row is the correct design when some type layers and some vector layers are selected. There should just be one swatch/blend-mode row, and the behavior should be intuitively uniform: the swatch is used to change the type AND vector colors simultaneously, as appropriate for the given selected layer. Why should the user care that these two swatches are wired up to different underlying actions? Instead it should be possible to change all the colors in a single operation.

DivyaPrabhakar commented:
I think @iwehrman is right. Initially I was thinking that the current approach made sense because the user cases for choosing multiple layers at a time (ranked by priority) is 1.) To make a property change to all selected layers 2.) To move the selected layers on the canvas. 

In the case that the user selects the layers to move them around, my initial hesitation was that we can't assume they want to apply a change to a property (like fill) to all the layers selected at once, so we need to provide the separated text color and vector fill options. But in that case, we already know the primary objective for the user is to change the placement of the layers, not change a property so it doesn't matter if we provide one or two different fill options anyway. Another thing I saw is that when the vector fill and text color are enabled, the stroke swatch is below the text. I think it should be paired with the vector fill swatch. 

I will try and implement the one swatch approach and add it as a separate commit to this branch so people can try both ways and we can decide from there.

chadrolfs commented:
sounds good @DivyaPrabhakar, let @ktaki and I know when the new branch is ready and we'll give it a once over.

ktaki commented:
@DivyaPrabhakar Toggle Fill seems not functional.

<img width="312" alt="screen shot 2016-03-04 at 9 55 26 am" src="https://cloud.githubusercontent.com/assets/2029597/13535500/73a45b06-e1ef-11e5-86ca-26f7d81ec045.png">

ktaki commented:
The Sampler is broken in this branch. Make sure this isn't a side-effect of your change.

chadrolfs commented:
Comments on new branch (mixed-appearance-refactoring-2):

Hitting Tab key after editing a field in mixed selection state results in an error:
![screen shot 2016-03-04 at 11 39 44 am](https://cloud.githubusercontent.com/assets/8442631/13538169/d848d510-e1fd-11e5-89cb-316345761d9d.png)

chadrolfs commented:
rotating a mixed selection changes the attributes but shouldn't
![mixed-rotate](https://cloud.githubusercontent.com/assets/8442631/13538315/b51ddd32-e1fe-11e5-9e55-7409e9263c59.gif)

DivyaPrabhakar commented:
@ktaki and @chadrolfs -- I have updated the PR description with all the bugs that this PR should fix -- mainly UI fixes. Also the problems that you both identified before should @ktaki and @chadrolfs -- The problems that you both identified before should be fixed: 

1.) Color swatch keeps the fill when the selections are rotated
![](https://s3.amazonaws.com/f.cl.ly/items/341P3S1Z142E362K0L2r/Screen%20Recording%202016-03-08%20at%2012.16%20PM.gif?v=6b046d95)

2.) Sampler tool works 
![](https://s3.amazonaws.com/f.cl.ly/items/260Z062t082p0i201r2M/Screen%20Recording%202016-03-08%20at%2012.19%20PM.gif?v=c88eb148)

3.) Hitting Tab key after editing Typeface field in mixed selection state results in an error:Hitting Tab key after editing Typeface field in mixed selection state results in no error now.

chadrolfs commented:
color picker is blinks mixed:
![blinky-cp](https://cloud.githubusercontent.com/assets/8442631/13617723/0100a4a8-e535-11e5-940f-75d750cf484a.gif)

chadrolfs commented:
interplay between the opacity in the color picker and the panel opacity is funky with a mixed selection that includes a pixel layer.

chadrolfs commented:
changing the tracking fails:
`Action type.setTracking failed: Error: Photoshop returned an error when playing command number: 1 Photoshop code: -25920 Photoshop message: error: -25920 message: The command “Set” is not currently available.`

chadrolfs commented:
same thing with Alignment:
`Action type.setAlignment failed: Error: Photoshop returned an error when playing command number: 1 Photoshop code: -25920 Photoshop message: error: -25920 message: The command “Set” is not currently available.`

ktaki commented:
The issues with manipulating alpha value could be caused by the issue reported in #3786 in which the alpha value for the text layers is treated differently from that of shape layers.

ktaki commented:
I did not find any other new issues.

DivyaPrabhakar commented on src/js/jsx/sections/style/AppearancePanelSections.jsx:
```diff
@@ -0,0 +1,491 @@
+                    if (characterStyles && !characterStyles.isEmpty()) {
+                        var textColors = collection.pluck(characterStyles, "color");
+                        var vectorColors = this.state.vectorFill.colors;
+                        var temp = textColors.concat(vectorColors);
```
@iwehrman Can you help me here? The problem that happens is that `textColors` is in the same format as `vectorColors` but apparently the `r, g, b, a,` values of `textColors` is not rounded to the nearest integer and `vectorColors` are. Should I just call `Math.round()` on each of these values or am I missing something if I do that? Just not sure because we clearly handle text colors and vector fills differently so I wanted to double check. @pineapplespatula You worked with type right? So maybe you know better? 

The end goal of this is getting all the vector fills and all the text colors, and putting them into one final data structure that is of the same type as `textColors`.

DivyaPrabhakar commented:
@chadrolfs I fixed the errors that were occurring on alignment and tracking and I fixed part of the opacity issue -- before if you used the slider, only text and vector layers would change opacity, but now all layers change opacity. Also the user can change opacity via the opacity input field i.e. type opacity = 30 and all the layers will adjust accordingly. 

2 known problems still exist -- 1.) When the opacity slider is adjusted, the opacity input field does not reflect this change. This is because of the open bug @ktaki mentioned above, and we have decided not to fix that bug. 2.) The color swatch overlay goes to mixed state when the user rotates the layers. I know why this is happening and was working on the fix but I need a little help from @iwehrman or @pineapplespatula, and I have commented on that line and will wait for their feedback. 

The rest of the issues both you and @ktaki found should have been addressed. I will let you know right when a decision on problem no. 2 is made. Sorry for the delay on this, it took me some time to narrow down the problem 2.

DivyaPrabhakar commented:
@chadrolfs The rotation error should be fixed now!

chadrolfs commented:
OK, I think this'll do it!  Thanks for all your hard work on this.  Ok to merge!

volfied commented on src/js/actions/sampler.js:
```diff
@@ -211,16 +211,12 @@ define(function (require, exports) {
      * @return {Promise}
      */
     var applyColor = function (doc, layers, color) {
```
so what's the point of passing layers to this now, if we're just gonna ignore it?

DivyaPrabhakar commented on src/js/models/color.js:
```diff
@@ -96,9 +96,9 @@ define(function (require, exports, module) {
-            "r": rgb.red,
-            "g": green,
-            "b": rgb.blue,
+            "r": mathjs.round(rgb.red),
```
FYI: This is because colors for type layers were returning as floating points and vector layers colors were integers. Created a problem when we called .uniformValue() on a group of type and vector colors, even though they were the same.

volfied commented:
Looks good to me, since it's also heavily tested, I saw no blaring errors.
